ui/packages/catalog-ui-search/src/main/webapp/js/widgets/cesium.base.line.js

```diff
@@ -0,0 +1,182 @@
+const wreqr = require('wreqr');
+const NotificationView = require('./notification.view');
+const DrawingController = require('./drawing.controller');
+
+"use strict";
+
+const CAMERA_MAGNITUDE_THRESHOLD = 8000000;
+
+const capitalize = (value) => value.charAt(0).toUpperCase() + value.slice(1);
+
+const getCurrentMagnitudeFromView = (view) => {
+    return view.map.camera.getMagnitude();
+}
+
+const needsRedraw = (view) => {
+    const currentMagnitude = getCurrentMagnitudeFromView(view);
+
+    if (view.cameraMagnitude < CAMERA_MAGNITUDE_THRESHOLD && currentMagnitude > CAMERA_MAGNITUDE_THRESHOLD) {
+        return true;
+    }
+    if (view.cameraMagnitude > CAMERA_MAGNITUDE_THRESHOLD && currentMagnitude < CAMERA_MAGNITUDE_THRESHOLD) {
+        return true;
+    }
+
+    return false;
+}
+
+class GeometryRenderView extends Marionette.View {
+  constructor(options) {
+    super(...options);
+
+    Object.assign(this, {...options});
+  }
+
+  updatePrimitive = () => {
+    this.drawGeometry(this.model);
+  }
+
+  drawGeometry = (model) => {
+    throw new Error('This method must be overwritten by implementing geometry!');
+  }
+
+  constructLinePrimitive = (coordinates) => {
+    const color = this.model.get('color');
+
+    return {
+      width: 8,
+      material: Cesium.Material.fromType('PolylineOutline', {
+          color: color ? Cesium.Color.fromCssColorString(color) : Cesium.Color.KHAKI,
+          outlineColor: Cesium.Color.WHITE,
+          outlineWidth: 4
+      }),
+      id: 'userDrawing',
+      positions: Cesium.Cartesian3.fromDegreesArray(_.flatten(coordinates))
+    };
+  }
+
+  listenForCameraChange = () => {
+    this.map.scene.camera.moveStart.addEventListener(this.handleCameraMoveStart, this);
+    this.map.scene.camera.moveEnd.addEventListener(this.handleCameraMoveEnd, this);
+  }
+
+  handleCameraMoveStart = () => {
+    this.animationFrameId = window.requestAnimationFrame(function() {
+        if (needsRedraw(this)){
+            this.updatePrimitive();
+        }
+        this.handleCameraMoveStart();
+    }.bind(this));
+  }
+
+  handleCameraMoveEnd = () => {
+    window.cancelAnimationFrame(this.animationFrameId);
+    if (needsRedraw(this)) {
+        this.updatePrimitive();
+    }
+  }
+
+  destroy = () => {
+    if (!this.map || !this.map.scene) {
+      return;
+    }
+
+    this.map.scene.camera.moveStart.removeEventListener(this.handleCameraMoveStart, this);
+    this.map.scene.camera.moveEnd.removeEventListener(this.handleCameraMoveEnd, this);
+    window.cancelAnimationFrame(this.animationFrameId);
+    if (this.primitive) {
+        this.map.scene.primitives.remove(this.primitive);
+    }
+    this.remove(); // backbone cleanup.
+  }
+}
+
+class GeometryController extends DrawingController {
+  constructor(options) {
+    DrawingController.prototype.drawingType = options.drawingType;
+    DrawingController.prototype.options = options;
+    
+    super(...options);
+
+    Object.assign(this, {...options});
```

Comment:
❓ If options is an object passed in, do we need to create a new object from it?

Reply:
See my other response.